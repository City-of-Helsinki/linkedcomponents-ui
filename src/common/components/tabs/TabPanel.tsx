import React from 'react';

export type TabPanelProps = React.PropsWithChildren<{
  className?: string;
  index?: number;
  isActive?: boolean;
  name?: string;
  style?: React.CSSProperties;
}>;

const TabPanel = ({
  children,
  className,
  index,
  isActive,
  name,
  style,
}: TabPanelProps) => {
  return isActive ? (
    <div
      id={`tab-${name}-${index}-panel`}
      role="tabpanel"
      aria-labelledby={`tab-${name}-${index}-button`}
      className={className}
      style={style}
    >
      {children}
    </div>
  ) : null;
};

TabPanel.componentName = 'TabPanel';

export default TabPanel;
