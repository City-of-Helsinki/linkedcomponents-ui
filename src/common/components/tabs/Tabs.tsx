import { ClassNames } from '@emotion/react';
import React from 'react';

import { useTheme } from '../../../domain/app/theme/Theme';
import { FCWithName, OptionType } from '../../../types';
import Tab from './tab/Tab';
import styles from './tabs.module.scss';

type TabOptionType = {
  isCompleted?: boolean;
} & OptionType;

interface Props {
  activeTab: string;
  className?: string;
  name: string;
  onChange: (selected: string) => void;
  options: TabOptionType[];
}

const Tabs: React.FC<React.PropsWithChildren<Props>> = ({
  activeTab,
  className,
  children,
  name,
  onChange,
  options,
}) => {
  const { theme } = useTheme();
  const isFocused = React.useRef(false);
  const activeIndex = React.useMemo(
    () => options.findIndex((option) => option.value === activeTab),
    [activeTab, options]
  );
  const [focusedTab, setFocusedTab] = React.useState(0);

  const handleChange = (newLanguage: string) => {
    onChange(newLanguage);
  };

  const onKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'ArrowRight') {
      setFocusedTab(Math.min(focusedTab + 1, options.length - 1));
    }
    if (event.key === 'ArrowLeft') {
      setFocusedTab(Math.max(focusedTab - 1, 0));
    }
  };

  const tabPanels = React.Children.toArray(children)
    .filter((child) => {
      return (
        React.isValidElement(child) &&
        (child.type as FCWithName).componentName === 'TabPanel'
      );
    })
    .map((child, index) => {
      return React.cloneElement(child as React.ReactElement, {
        index,
        name,
        isActive: activeIndex === index,
      });
    });

  return (
    <ClassNames>
      {({ css, cx }) => (
        <div
          className={className}
          onBlur={() => {
            isFocused.current = false;
          }}
          onFocus={() => {
            isFocused.current = true;
          }}
        >
          <div
            className={cx(styles.tabList, css(theme.tabs))}
            role="tablist"
            onKeyDown={onKeyDown}
          >
            {options.map((option, index) => {
              return (
                <Tab
                  key={option.value}
                  autoFocusChange={isFocused.current}
                  index={index}
                  isCompleted={option.isCompleted}
                  isFocused={index === focusedTab}
                  isActive={option.value === activeTab}
                  name={name}
                  option={option}
                  onClick={handleChange}
                  setFocusedTab={setFocusedTab}
                />
              );
            })}
          </div>
          {tabPanels}
        </div>
      )}
    </ClassNames>
  );
};

export default Tabs;
