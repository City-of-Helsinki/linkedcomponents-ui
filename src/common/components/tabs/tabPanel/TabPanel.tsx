import React from 'react';

import { FCWithName } from '../../../../types';

export type TabPanelProps = React.PropsWithChildren<{
  className?: string;
  index?: number;
  isActive?: boolean;
  name?: string;
  style?: React.CSSProperties;
}>;

const TabPanel: FCWithName<TabPanelProps> = ({
  children,
  className,
  index,
  isActive,
  name,
  style,
}) => {
  return isActive ? (
    <div
      id={`tab-${name}-${index}-panel`}
      role="tabpanel"
      aria-labelledby={`tab-${name}-${index}-button`}
      className={className}
      style={style}
    >
      {children}
    </div>
  ) : null;
};

TabPanel.componentName = 'TabPanel';

export default TabPanel;
