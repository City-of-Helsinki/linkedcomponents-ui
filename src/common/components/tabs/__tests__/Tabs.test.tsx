import React from 'react';

import {
  arrowLeftKeyPressHelper,
  arrowRightKeyPressHelper,
  configure,
  render,
  screen,
  userEvent,
} from '../../../../utils/testUtils';
import TabPanel from '../tabPanel/TabPanel';
import Tabs from '../Tabs';

configure({ defaultHidden: true });

const options = [
  {
    isCompleted: true,
    label: 'Tab 1',
    value: 'tab1',
  },
  {
    isCompleted: false,
    label: 'Tab 2',
    value: 'tab2',
  },
  {
    isCompleted: false,
    label: 'Tab 3',
    value: 'tab3',
  },
];

const getElement = (
  key: 'panel1' | 'panel2' | 'panel3' | 'tab1' | 'tab2' | 'tab3'
) => {
  switch (key) {
    case 'panel1':
      return screen.getByRole('tabpanel', { name: 'Tab 1' });
    case 'panel2':
      return screen.getByRole('tabpanel', { name: 'Tab 2' });
    case 'panel3':
      return screen.getByRole('tabpanel', { name: 'Tab 2' });
    case 'tab1':
      return screen.getByRole('tab', { name: 'Tab 1' });
    case 'tab2':
      return screen.getByRole('tab', { name: 'Tab 2' });
    case 'tab3':
      return screen.getByRole('tab', { name: 'Tab 3' });
  }
};

const renderComponent = (onChange = jest.fn()) =>
  render(
    <Tabs activeTab="tab1" name="tabs" onChange={onChange} options={options}>
      <TabPanel>Panel 1</TabPanel>
      <TabPanel>Panel 2</TabPanel>
      <TabPanel>Panel 3</TabPanel>
    </Tabs>
  );

test('renders the component', () => {
  const { container } = renderComponent();

  getElement('panel1');
  getElement('tab1');
  getElement('tab2');
  getElement('tab3');

  expect(
    screen.queryByRole('tabpanel', { name: 'Tab 2' })
  ).not.toBeInTheDocument();
  expect(
    screen.queryByRole('tabpanel', { name: 'Tab 3' })
  ).not.toBeInTheDocument();

  expect(container).toMatchSnapshot();
});

test('should call onKeyDown and change focused tab', async () => {
  const user = userEvent.setup();
  renderComponent();

  const tab1 = getElement('tab1');
  const tab2 = getElement('tab2');
  const tab3 = getElement('tab3');
  await user.click(tab1);
  expect(tab1).toHaveFocus();

  arrowRightKeyPressHelper(tab1);
  expect(tab2).toHaveFocus();
  arrowRightKeyPressHelper(tab2);
  expect(tab3).toHaveFocus();
  arrowLeftKeyPressHelper(tab3);
  expect(tab2).toHaveFocus();
});

test('should call onChange', async () => {
  const onChange = jest.fn();
  const user = userEvent.setup();
  renderComponent(onChange);

  const tab1 = getElement('tab1');
  const tab2 = getElement('tab2');
  await user.click(tab1);
  expect(tab1).toHaveFocus();

  await user.click(tab2);
  expect(tab2).toHaveFocus();
  expect(onChange).toBeCalledWith('tab2');
});
