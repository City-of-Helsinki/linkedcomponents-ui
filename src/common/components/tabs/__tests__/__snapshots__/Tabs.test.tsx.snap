// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the component 1`] = `
<div>
  <div>
    <div
      class="tabList css-0"
      role="tablist"
    >
      <button
        aria-controls="tab-tabs-0-panel"
        aria-selected="true"
        class="tab isActive"
        id="tab-tabs-0-button"
        role="tab"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          aria-label="check"
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_small__gTGkU icon_hds-icon--size-s__2Lkik icon"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M21 7L10 18L4.5 12.5L6 11L10 15L19.5 5.5L21 7Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        Tab 1
      </button>
      <button
        aria-controls="tab-tabs-1-panel"
        aria-selected="false"
        class="tab"
        id="tab-tabs-1-button"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 2
      </button>
      <button
        aria-controls="tab-tabs-2-panel"
        aria-selected="false"
        class="tab"
        id="tab-tabs-2-button"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 3
      </button>
    </div>
    <div
      aria-labelledby="tab-tabs-0-button"
      id="tab-tabs-0-panel"
      role="tabpanel"
    >
      Panel 1
    </div>
  </div>
</div>
`;
